### PR TITLE
[RFR] Deck and card management

### PR DESCRIPTION
**Pull request by Kmaschta**

### CODE REVIEW

File: tests/DobbleCardTest.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            [[1, 2, 3]],
+            [['A', 'B', 'C']],
+            [['heart', 'square', 'diamond', 'club']],
+        ];
+    }
+
+    /**
+     * @dataProvider validIterables
+     */
+    public function testCardAcceptIterable($iterable)
```
Comment (fzaninotto) [inline tests/DobbleCardTest.php]: s/testCardAcceptIterable/testCardConstructorAcceptsIterable

same for the next method

File: src/Dobble/Card.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+namespace Marmelab\Dobble;
+
+use Countable;
+
+class Card implements Countable
```
Comment (jpetitcolas) [inline src/Dobble/Card.php]: Shouldn't it be `\Countable`?

File: src/Dobble/Deck.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+namespace Marmelab\Dobble;
+
+/**
+ * @todo validate & generate functions
+ */
+class Deck implements \Countable
+{
+    private $cards = [];
+
+    public function __construct($cards = [])
```
Comment (jpetitcolas) [inline src/Dobble/Deck.php]: Typehint

File: src/Dobble/Deck.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        if (!is_array($cards) && !$cards instanceof \Traversable) {
```
Comment (jpetitcolas) [inline src/Dobble/Deck.php]: With typehinting, you can remove these checks.

File: src/Dobble/Deck.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            throw new \InvalidArgumentException('Not a list of cards');
+        }
+
+        foreach ($cards as $card) {
+            $this->append($card);
```
Comment (jpetitcolas) [inline src/Dobble/Deck.php:15]: You currently have no specific behavior in your `append` method. Keep it simple by just setting `$this->cards = $cards`.

File: src/Dobble/Deck.php (hunk 4 above)
Comment (Kmaschta) [inline src/Dobble/Deck.php]: I'm ok for this use case, but what is the solution if we want to pass all types of iterable ?
An array is sadly not a [Traversable](http://php.net/Traversable) but all other objects which can pass into a foreach are.

File: src/Dobble/Deck.php (hunk 5 above)
Comment (Kmaschta) [inline src/Dobble/Deck.php:15]: How check if all elements in the array are instance of Card class ?

File: src/Dobble/Card.php [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    private $symbols = [];
+
+    public function __construct(array $symbols)
+    {
+        $this->symbols = $symbols;
+
+        // The deck validation uses the fonction "array_unique" to check
+        // if some cards have same values, so we need to sort symbols
+        // for avoid that same values seems differents.
+        asort($this->symbols);
```
Comment (jpetitcolas) [inline src/Dobble/Card.php:38]: So, why isn't it the deck who sort card symbols?

File: src/Dobble/Deck.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    public function validate()
+    {
+        if (!count($this)) {
+            // The deck is empty
+            return false;
+        }
+
+        if (count(array_unique($this->cards)) < count($this)) {
+            // The deck has two or more identical cards
+            return false;
```
Comment (jpetitcolas) [inline src/Dobble/Deck.php]: I would instead throw exceptions, to give more details on why a deck is not valid.

File: src/Dobble/Deck.php (hunk 5 above)
Comment (jpetitcolas) [inline src/Dobble/Deck.php:15]: Good point!

**Comment by jpetitcolas** [comment]
:+1:

File: src/Dobble/Card.php (hunk 6 above)
Comment (Kmaschta) [inline src/Dobble/Card.php:38]: The deck need cards with sorted symbols but the cards can be in this state before, I think it's more elegant to display an array which is already sorted.
However, the `$symbols` attribute is private.

Is it a good reason or there is a better solution ?

File: src/Dobble/Deck.php (hunk 7 above)
Comment (Kmaschta) [inline src/Dobble/Deck.php]: :+1:, next commit.

File: .gitignore [hunk 8]
```diff
@@ -1 +1,2 @@
 /vendor
+*.vim
```
Comment (jpetitcolas) [inline .gitignore]: Move it into your `.git/info/exclude` file as it is not project specific.

Comment (Kmaschta) [inline .gitignore]: Done

**Pull request merged**